cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/OperationsDashboardHttpHandler.java

```diff
@@ -0,0 +1,133 @@
+import co.cask.cdap.proto.NamespaceMeta;
+import co.cask.cdap.proto.id.NamespaceId;
+import co.cask.cdap.proto.id.ProgramRunId;
+import co.cask.cdap.proto.ops.DashboardProgramRunRecord;
+import co.cask.http.HttpResponder;
+import com.google.gson.Gson;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+import io.netty.handler.codec.http.FullHttpRequest;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.QueryParam;
+
+/**
+ * {@link co.cask.http.HttpHandler} to handle program run operation dashboard and reports for v3 REST APIs
+ */
+@Singleton
+@Path(Constants.Gateway.API_VERSION_3)
+public class OperationsDashboardHttpHandler extends AbstractAppFabricHttpHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(OperationsDashboardHttpHandler.class);
+  private static final Gson GSON = new Gson();
+  private final Store store;
+  private final NamespaceAdmin namespaceAdmin;
+
+  @Inject
+  public OperationsDashboardHttpHandler(Store store, NamespaceAdmin namespaceAdmin) {
+    this.store = store;
+    this.namespaceAdmin = namespaceAdmin;
+  }
+
+  // TODO: [CDAP-13351] Need to support getting scheduled program runs if start + duration is ahead of the current time
+  @GET
+  @Path("/dashboard")
+  public void readDashboardDetail(FullHttpRequest request, HttpResponder responder,
+                                  @QueryParam("start") long startTimeSecs,
+                                  @QueryParam("duration") int durationTimeSecs,
+                                  @QueryParam("namespace") Set<String> namespaces)
+    throws IOException, BadRequestException {
+    if (startTimeSecs < 0) {
+      responder.sendString(HttpResponseStatus.BAD_REQUEST, "'start' time cannot be smaller than 0.");
+      return;
+    }
+    if (durationTimeSecs < 0) {
+      responder.sendString(HttpResponseStatus.BAD_REQUEST, "'duration' cannot be smaller than 0.");
+      return;
+    }
+    Set<String> existingNamespaces;
+    try {
+      existingNamespaces = namespaceAdmin.list().stream().map(NamespaceMeta::getName).collect(Collectors.toSet());
+      List<String> invalidNamespaces =
+        namespaces.stream().filter(ns -> !existingNamespaces.contains(ns)).collect(Collectors.toList());
+      // if there are non-existent namespaces in the request, respond with bad request status
+      if (!invalidNamespaces.isEmpty()) {
+        responder.sendString(HttpResponseStatus.BAD_REQUEST,
+                             "Request contains invalid namespaces: " + invalidNamespaces.toString());
+        return;
+      }
+    } catch (Exception e) {
+      LOG.error("Failed to list namespaces.", e);
```

Comment:
So just log and then continue? Is that correct?

Reply:
on a second thought, this check seems not really useful and complicates the logic. Maybe I can remove this?

Reply:
If it is not a failure, then why bother checking? I think you can remove it.